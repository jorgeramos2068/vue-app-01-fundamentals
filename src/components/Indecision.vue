<template>
  <img v-if="image" :src="image" alt="bg" />
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <input type="text" placeholder="Ask me a question" v-model="question" />
    <p>Remember to finish with a ?</p>
    <div v-if="isValidQuestion">
      <h2>{{ question }}</h2>
      <h1>{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  name: 'Indecision',
  data() {
    return {
      question: null,
      answer: null,
      image: null,
      isValidQuestion: false,
    };
  },
  methods: {
    async getAnswer() {
      try {
        this.answer = 'Loading...';
        const { answer, image } = await fetch('https://yesno.wtf/api').then(
          resp => resp.json()
        );
        this.answer = answer;
        this.image = image;
      } catch (err) {
        console.error(err);
      }
    },
  },
  watch: {
    question(value, oldValue) {
      this.isValidQuestion = false;
      if (!value.endsWith('?')) {
        return;
      }
      this.isValidQuestion = true;
      this.getAnswer();
    },
  },
};
</script>

<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
}

input {
  border: none;
  border-radius: 5px;
  margin-bottom: 10px;
  padding: 10px 15px;
  width: 250px;
}
input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>
