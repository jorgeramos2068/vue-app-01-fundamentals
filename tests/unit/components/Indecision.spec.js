import { shallowMount } from '@vue/test-utils';
import Indecision from '@/components/Indecision';

describe('Indecision component', () => {
  let wrapper;
  let clgSpy;

  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: 'yes',
          forced: false,
          image: 'https://yesno.wtf/assets/yes/2.gif',
        }),
    })
  );

  beforeEach(() => {
    wrapper = shallowMount(Indecision);
    clgSpy = jest.spyOn(console, 'log');
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should not call the getAnswer method when writting in the input without "?"', async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer');
    const input = wrapper.find('input');
    await input.setValue('Test value');
    expect(clgSpy).toHaveBeenCalledTimes(1);
    expect(getAnswerSpy).not.toHaveBeenCalled();
  });

  it('should call the getAnswer method when writting the "?" character in the input', async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer');
    const input = wrapper.find('input');
    await input.setValue('Test value?');
    expect(clgSpy).toHaveBeenCalledTimes(1);
    expect(getAnswerSpy).toHaveBeenCalled();
  });

  it('should call the getAnswer() method with a successful response', async () => {
    await wrapper.vm.getAnswer();
    expect(wrapper.vm.answer).toBe('yes');
    expect(wrapper.vm.image).toBe('https://yesno.wtf/assets/yes/2.gif');
    const img = wrapper.find('img');
    expect(img.exists()).toBeTruthy();
  });

  it('should call the getAnswer() method with an errored response', async () => {
    fetch.mockImplementationOnce(() => Promise.reject('API is down'));
    await wrapper.vm.getAnswer();
    const img = wrapper.find('img');
    expect(img.exists()).toBeFalsy();
    expect(wrapper.vm.answer).toBe('Error while loading...');
    expect(wrapper.vm.image).toBeNull();
  });
});
