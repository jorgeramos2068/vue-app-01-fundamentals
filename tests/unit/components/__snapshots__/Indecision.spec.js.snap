// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision component should match the snapshot 1`] = `
<div>
  <!--v-if-->
  <div class="bg-dark"></div>
  <div class="indecision-container"><input type="text" placeholder="Ask me a question">
    <p>Remember to finish with a ?</p>
    <!--v-if-->
  </div>
</div>
`;
